### ISSUE TITLE
dotnet exe under /usr/lib64 should have bin_t context

### ISSUE THREAD

**Issue by tmds** [fedora-selinux/selinux-policy#259]
Hi, .NET Core packages from [Fedora .NET SIG](https://fedoraproject.org/wiki/SIGs/DotNet) install an executable at `/usr/lib64/dotnet/dotnet`. A symlink is added to `/usr/bin/dotnet` which links to that executable.

The `dotnet` executable isn't usable from a systemd system service: SELinux blocks it because it has the `lib_t` context.

Changing to `bin_t` makes it work:
```
sudo semanage fcontext -a -t bin_t /usr/lib/dotnet/dotnet
sudo restorecon -R -v /usr/lib64/dotnet/dotnet
```

Can you please update the Fedora SELinux policy so the executable gets the `bin_t` label by default?

CC @omajid @RheaAyase

**Comment by ghost** [comment]
Tom Deseyn <notifications@github.com> writes:

> Hi, .NET Core packages from Fedora .NET SIG install an executable at /usr/lib64/dotnet/dotnet. A symlink is added to /usr/bin/dotnet which
> links to that executable.
>
> The dotnet executable isn't usable from a systemd system service: SELinux blocks it because it has the lib_t context.
>
> Changing to bin_t makes it work:
>
> sudo semanage fcontext -a -t bin_t /usr/lib/dotnet/dotnet
> sudo restorecon -R -v /usr/lib64/dotnet/dotnet
>
> Can you please update the Fedora SELinux policy so the executable gets the bin_t label by default?

Not my call but would it not instead be better to fix the packaging?

>
> CC @omajid @RheaAyase
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

-- 
Key fingerprint = 5F4D 3CDB D3F8 3652 FBD8 02D5 3B6C 5F1D 2C7B 6B02
https://sks-keyservers.net/pks/lookup?op=get&search=0x3B6C5F1D2C7B6B02
Dominick Grift

**Comment by RheaAyase** [comment]
TLDR would be "The packaging is as intended and correct."

Long story: https://docs.microsoft.com/en-us/dotnet/core/build/distribution-packaging

**Comment by tmds** [comment]
> Not my call but would it not instead be better to fix the packaging?

We can't move the `dotnet` exe because it needs to be at the 'root' of a hierarchy of folders. [.NET Core distribution packaging](https://docs.microsoft.com/en-us/dotnet/core/build/distribution-packaging) describes the layout. Is it that what you mean? Or maybe you have something else in mind?

**Comment by RheaAyase** [comment]
That comment synchronization! \o/ :D
(please ignore me)

**Comment by ghost** [comment]
Thanks. duly noted. I was just trying to consolidate that with this: https://refspecs.linuxfoundation.org/FHS_3.0/fhs/ch04s06.html

**Comment by RheaAyase** [comment]
It is definitely a good point and perhaps we should bring it up with MS. I don't even remember anymore what was the reason for this to be there, it may not exist anymore after other kinds of changes. @tmds @omajid what do you think?

**Comment by tmds** [comment]
I'd rather get this to work than start a discussion with Microsoft.
They'll want to keep the exe at the root because it makes it simple to pack things up in a tar.gz file.
If we'd place it elsewhere and make it work, we can't be sure it will keep on working in the future.

**Comment by omajid** [comment]
@doverride .NET placing a binary here is a bit unfortunate (I would have preferred FHS too), but it is not too different from java packages placing binaries at `/usr/lib/..../java` in Fedora.

**Comment by tmds** [comment]
@doverride can you look into this please? or assign it to someone that can?

**Comment by ghost** [comment]
Tom Deseyn <notifications@github.com> writes:

> @doverride can you look into this please? or assign it to someone that can?

@wrabcak can you add this spec?

>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

-- 
Key fingerprint = 5F4D 3CDB D3F8 3652 FBD8 02D5 3B6C 5F1D 2C7B 6B02
https://sks-keyservers.net/pks/lookup?op=get&search=0x3B6C5F1D2C7B6B02
Dominick Grift

**Comment by zpytela** [comment]
I can do that.

**Comment by ghost** [comment]
Zdeněk Pytela <notifications@github.com> writes:

> I can do that.

Thanks

>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

-- 
Key fingerprint = 5F4D 3CDB D3F8 3652 FBD8 02D5 3B6C 5F1D 2C7B 6B02
https://sks-keyservers.net/pks/lookup?op=get&search=0x3B6C5F1D2C7B6B02
Dominick Grift

**Comment by zpytela** [comment]
I've just noticed 2 different path specifications:

/usr/lib/dotnet/dotnet
/usr/lib64/dotnet/dotnet

Which one is correct?

**Comment by ghost** [comment]
Zdeněk Pytela <notifications@github.com> writes:

> I've just noticed 2 different path specifications:
>
> /usr/lib/dotnet/dotnet
> /usr/lib64/dotnet/dotnet
>
> Which one is correct?

/usr/lib/dotnet/dotnet

see: https://github.com/fedora-selinux/selinux-policy/blob/rawhide/config/file_contexts.subs_dist#L17

>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

-- 
Key fingerprint = 5F4D 3CDB D3F8 3652 FBD8 02D5 3B6C 5F1D 2C7B 6B02
https://sks-keyservers.net/pks/lookup?op=get&search=0x3B6C5F1D2C7B6B02
Dominick Grift

**Comment by zpytela** [comment]
@doverride thank you for all you comments. My main question is how to find it out - when the rule is to follow the module/package name or type like in this case.
I understand that declaration in a particular module is quite a strong indication, so asking more generally.

**Comment by tmds** [comment]
@zpytela the way we figured this out is: the dotnet exe is at /usr/lib64/dotnet/dotnet, but when you pass that to semanage, it gives an error:

> ValueError: File spec /usr/lib64/dotnet/dotnet conflicts with equivalency rule '/usr/lib64 /usr/lib'; Try adding '/usr/lib/dotnet/dotnet' instead

**Comment by zpytela** [comment]
@tmds it is because of equivalency fcontext rules, see tail of semanage-fcontext:
```
semanage fcontext -l | tail -n 20
/lib = /usr/lib
/lib64 = /usr/lib
/usr/lib64 = /usr/lib
/usr/local/lib64 = /usr/lib
/usr/local/lib32 = /usr/lib
...
```
The rules need to be set upon source path of the equivalency rule.

**Comment by ghost** [comment]
Zdeněk Pytela <notifications@github.com> writes:

> @doverride thank you for all you comments. My main question is how to find it out - when the rule is
> to follow the module/package name or type like in this case.
> I understand that declaration in a particular module is quite a strong indication, so asking more
> generally.

Good question, and I do not have a good answer for this other than: if
you consistently add specs to modules that declares the type referenced
in the spec then you should generally be ok.

If you want to debate this topic or get an authoritative answer to your
question then you should probably take this to the reference policy
maillist so that the upstream maintainer can explain.

>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

-- 
Key fingerprint = 5F4D 3CDB D3F8 3652 FBD8 02D5 3B6C 5F1D 2C7B 6B02
https://sks-keyservers.net/pks/lookup?op=get&search=0x3B6C5F1D2C7B6B02
Dominick Grift

**Comment by ghost** [comment]
It has been a long time since i've had hands-on experience with reference policy-based module policy. The details have lost, but if i recall correctly: you have to look at this in the bigger picture of the modularity aspect of it all. If you reference a type in a module then that module depends on the module where there referenced type is declared. So in your case where you reference type bin_t in the miscfiles.fc file. youre adding a dependency for miscfiles on corecommands. I this case that is not a big deal since the corecommands is part of base and and so you can rely on it to alway's be there. However depending on the scenario that might not always be the case. That is why i think one should always consistently add specs to the module where the referenced type is declared. It is a simple consistency rule that make things easy not to forget. It also has practical benefits in that if you consistently follow this pattern then the policy becomes more readible/discoverable since you can use the simple rule to determine where a specific spec is located.

**Comment by ghost** [comment]
Point is that it is module policy. If you treat it as a monolithic 'blob' then you eventually might lose some of the benefits of modularity.

**Comment by zpytela** [comment]
Understood. Although in this case it does not completely apply as you can hardly remove a module like corecommands.

**Comment by ghost** [comment]
Exactly, but consistency is important too.

**Comment by tmds** [comment]
@zpytela @doverride Thanks!
